pkg/proctree/nodes/process.go

```diff
@@ -0,0 +1,239 @@
+	interpreter *FileInfo      // current task interpreter info
+	exitTime    uint64         // current task exit time (if exited)
+	children    map[uint32]any // hashes for parents children (unique identifiers)
+	isThread    bool           // is this a lightweight process ?
+	mutex       *sync.RWMutex  // synchronize this process node
+}
+
+func NewProcess(hash uint32) *Process {
+	return &Process{
+		current:     hash,
+		parent:      0,
+		info:        &ProcessInfo{},
+		command:     &FileInfo{},
+		interpreter: &FileInfo{},
+		exitTime:    0,
+		children:    make(map[uint32]any),
+		isThread:    false,
+		mutex:       &sync.RWMutex{},
+	}
+}
+
+//
+// Setters (arguments are always immutable)
+//
+
+func (p *Process) SetParentHash(hash uint32) {
+	if hash == 0 {
+		return
+	}
+	p.mutex.Lock()
+	p.parent = hash
+	p.mutex.Unlock()
+}
+
+func (p *Process) AddChildHash(hash uint32) {
+	if hash == 0 {
+		return
+	}
+	p.mutex.Lock()
+	p.children[hash] = true
+	p.mutex.Unlock()
+}
+
+// SetInfo sets the parent and current task basic info.
+func (p *Process) SetInfo(info ProcessInfo) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	// Only sets the fields that aren't zero
+	if info.Tid > 0 {
+		p.info.Tid = info.Tid
+	}
+	if info.NsTid > 0 {
+		p.info.NsTid = info.NsTid
+	}
+	if info.Pid > 0 {
+		p.info.Pid = info.Pid
+	}
+	if info.NsPid > 0 {
+		p.info.NsPid = info.NsPid
+	}
+	if info.StartTime > 0 {
+		p.info.StartTime = info.StartTime
+	}
+	if info.ExitTime > 0 {
+		p.exitTime = info.ExitTime
+	}
+	// Derived fields
+	if info.Tid > 0 && info.Pid > 0 {
+		if info.Tid != info.Pid {
+			p.isThread = true
+		}
+	}
+}
+
+// SetCommandAndInterpreter sets the current task command and interpreter info.
+func (p *Process) SetCommandAndInterpreter(command, interpreter FileInfo) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.setFileInfo(0, command)
+	p.setFileInfo(1, interpreter)
+}
+
+// SetCommand sets the current task command info.
+func (p *Process) SetCommand(command FileInfo) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.setFileInfo(0, command)
+}
+
+// SetInterpreter sets the current task interpreter info.
+func (p *Process) SetInterpreter(interpreter FileInfo) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.setFileInfo(1, interpreter)
+}
+
+// setFileInfo sets the current task command or interpreter info (no lock).
+func (p *Process) setFileInfo(which int, info FileInfo) {
+	var ptr *FileInfo
+
+	switch which {
+	case 0:
+		ptr = p.command
+	case 1:
+		ptr = p.interpreter
+	}
+
+	// Only sets the fields that aren't zero
+
+	if info.Name != "" {
+		ptr.Name = info.Name
+	}
+	if info.Path != "" {
+		ptr.Path = info.Path
+	}
+	if info.Device > 0 {
+		ptr.Device = info.Device
+	}
+	if info.Ctime > 0 {
+		ptr.Ctime = info.Ctime
+	}
+	if info.Inode > 0 {
+		ptr.Inode = info.Inode
+	}
+	if info.InodeMode > 0 {
+		ptr.InodeMode = info.InodeMode
+	}
+}
+
+//
+// Getters (make sure to always return a copy)
```

Comment:
Use the Info types to always return immutable (non instantiable) data to callers.